csharp/ql/lib/semmle/code/csharp/Type.qll

```diff
@@ -704,15 +709,38 @@ class Enum extends ValueType, @enum_type {
  * ```
  */
 class Struct extends ValueType, @struct_type {
-  /** Holds if this `struct` has a `ref` modifier. */
-  predicate isRef() { this.hasModifier("ref") }
+  /**
+   * DEPRECATED: Use `instanceof RefStruct` instead.
+   *
+   * Holds if this `struct` has a `ref` modifier.
+   */
+  deprecated predicate isRef() { this.hasModifier("ref") }
 
   /** Holds if this `struct` has a `readonly` modifier. */
   predicate isReadonly() { this.hasModifier("readonly") }
 
   override string getAPrimaryQlClass() { result = "Struct" }
 }
 
+/**
+ * A `ref struct`, for example
+ *
+ * ```csharp
+ * ref struct S {
+ *  ...
+ * }
+ * ```
+ */
+class RefStruct extends Struct {
+  RefStruct() { this.hasModifier("ref") }
+
+  override string getAPrimaryQlClass() { result = "RefStruct" }
+
+  override predicate isValueType() { none() }
```

Comment:
Why don't we consider this a value type? 
```
using System;
void M(C c) 
{
    c.F = 5;
}

var c = new C();
M(c);                   // Passed by value
Console.WriteLine(c.F); // So prints 0

ref struct C 
{
    public int F;
}

```

Reply:
This is definitely up for debate.
As you write, the `ref struct` still exhibits call-by-value semantics. However, we also decided a while back that we needed to introduce post update nodes for ref structs parameters (as `ref structs` can have `ref` fields - and in this case only the reference is copied, which from a dataflow perspective means that we "can" update a ref struct (or at least update the content of the shared references)). Furthermore, `ref structs` can't be boxed (implicitly converted) to `ValueType` (or `object`). My thinking is that it is "something in between" and "reference like" (this is also the terminology used in Roslyn). So I thought that it was starting to get a bit dangerous to assume that it behaves like a value type both from a type and data flow perspective.

Reply:
To continue the discussion - I don't mind that we still consider the `ref struct` a value type - then we can modify type conversion, unification and dataflow on a case by case when we run into problems.
Maybe that is the safer choice.